### ISSUE TITLE
Keep-Alive and HTTP2 pings

### ISSUE THREAD

**Issue by SebastianThiebaud** [grpc/grpc-swift#843]
### What are you trying to achieve? 

The `grpc-swift` library built on top of the C library (0.x) used to expose a few channel arguments like `http2MinSentPingInterval` or `keepAliveTimeout`
https://github.com/grpc/grpc-swift/blob/cgrpc/Sources/SwiftGRPC/Core/ChannelArgument.swift#L46

I'm struggling to find the equivalent for the new library (1.x) and based on initial investigation, it doesn't seem that SwiftNIOHTTP2 implements any logic to ping. Am I missing something?

**Comment by Lukasa** [comment]
Right now the new library does not auto-ping, but it could! The infrastructure is there for it to do so, the code just isn't written yet.

**Comment by SebastianThiebaud** [comment]
@Lukasa Looking quickly at the codebase and how Swift NIO works (I'm still not 100% familiar with it), it seems like the best strategy would be to add a new handler (or even two: one for client and one for server) to the existing channel's pipeline that will handle keep alive (ping and pong + logic documented [here](https://github.com/grpc/grpc/blob/master/doc/keepalive.md)).

Is that correct? If so, I can start looking at it and maybe submitting a pull request in the near future.

**Comment by Lukasa** [comment]
Yeah, that would work. To implement the `WITHOUT_DATA` versions is a bit trickier and requires that the handler live in a very specific spot in the pipeline (between the `HTTP2Handler` and the `HTTP2StreamMultiplexer`).

**Comment by SebastianThiebaud** [comment]
I don't think I encountered `WITHOUT_DATA` yet. Is that the same thing as `GRPC_ARG_KEEPALIVE_PERMIT_WITHOUT_CALLS` aka no active RPCs?

**Comment by SebastianThiebaud** [comment]
Messing with it quickly, that's how I added the handlers to the pipelines.

```
                        [I] ↓↑ [O]
            NIOHTTP2Handler ↓↑ NIOHTTP2Handler            [handler0]
            GRPCIdleHandler ↓↑                            [handler3]
 GRPCClientKeepaliveHandler ↓↑ GRPCClientKeepaliveHandler [handler2]
     HTTP2StreamMultiplexer ↓↑ HTTP2StreamMultiplexer     [handler1]

                        [I] ↓↑ [O]
 GRPCServerKeepaliveHandler ↓↑ GRPCServerKeepaliveHandler [handler0]
       HTTPProtocolSwitcher ↓↑                            [handler1]
```

**Comment by Lukasa** [comment]
Yeah, except I think it's no actual headers/data frames moving on the wire.

Your pipelines look right to me.

I have the day off today as it's Juneteenth, but I'll take a review of your draft patch on Monday if that's ok.

**Comment by SebastianThiebaud** [comment]
Closing this issue as #850 is in review.
